### PR TITLE
CODAP-1328: accept date axes in univariate and count adornments

### PR DESCRIPTION
**Pull request by bfinzer**
## Summary

Follow-up to CODAP-1320 / PR #2574. Fixes four adornments that
over-restrictively checked for "numeric" attributes/axes and so
malfunctioned on date axes:

- **Box plot**: orientation + attrId both depended on `xAttrType === "numeric"`.
- **Normal curve**: model picked the wrong attribute for mean/stddev.
- **Movable value**: orientation and drag math depended on `xAttrType === "numeric"`.
- **Count**: `respondToAxisDomainChange` reaction never fired on date axes
  (gated by `isNumericAxisModel`).

Two additional fixes uncovered while testing on a date-attribute dot plot:

- Movable value label rendered raw seconds-since-epoch on a date axis
  (now passes `isDate` + `DateAxisModel.precisionForDisplay` to
  `formatValueForScale`; widened label CSS so dates fit on one line).
- Per-cell count text went stale on selection changes when
  "Show measures for selection" was on — the component had no
  selection-related observable in its render body, so no re-render
  fired (`useEffect` runs every render but is not reactive). Now
  observes `dataConfig.selection.length` in render.

Fixes CODAP-1328.

## Test plan

- [x] Lint, TS check, and Jest tests on the affected adornment files pass locally.
- [ ] Open a dot plot with a date attribute on x. Verify:
  - [ ] Box plot: vertical, statistics computed from the date attribute, IQR boxes positioned along the date axis.
  - [ ] Normal curve: vertical, mean/stddev derived from the date attribute.
  - [ ] Movable value: vertical, drag updates x value, label formatted as a date on one line.
  - [ ] Count adornment: text positions refresh on date-axis pan/zoom; per-cell counts refresh dynamically when selection changes with "Show measures for selection" on.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: v3/src/components/graph/adornments/univariate-measures/normal-curve/normal-curve-adornment-model.ts [hunk 1]
```diff
@@ -129,7 +130,7 @@ type CurveParamInstance = {
     updateCategories(options: IUpdateCategoriesOptions) {
       const { dataConfig, resetPoints } = options
       const { xAttrId, yAttrId, xAttrType } = dataConfig.getCategoriesOptions()
-      const attrId = xAttrId && xAttrType === "numeric" ? xAttrId : yAttrId
+      const attrId = xAttrId && isNumericAttributeType(xAttrType) ? xAttrId : yAttrId
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/adornments/univariate-measures/normal-curve/normal-curve-adornment-model.ts:133]: Agreed — real gap. Enabling date x-attributes here means the mean/stddev labels still flow through `UnivariateMeasureAdornmentHelper.formatValueForScale`, which calls `MultiScale.formatValueForScale` with `isDate=false`, so they'll render as seconds-since-epoch on a date axis. Leaving this open as a requested change — the helper needs the same `isDate` + precision treatment movable-value labels got.

File: v3/src/components/graph/adornments/univariate-measures/box-plot/box-plot-adornment-component.tsx [hunk 2]
```diff
@@ -42,12 +43,12 @@ export const BoxPlotAdornmentComponent = observer(function BoxPlotAdornmentCompo
...
   const dataConfig = useGraphDataConfigurationContext()
   const { xAttrId, yAttrId, xAttrType } = useAdornmentAttributes()
-  const isVerticalRef = useRef(!!(xAttrType && xAttrType === "numeric"))
+  const isVerticalRef = useRef(isNumericAttributeType(xAttrType))
   const { cellCounts } = useAdornmentCells(model, cellKey)
   const helper = useMemo(() => {
     return new UnivariateMeasureAdornmentHelper(cellKey, isVerticalRef, layout, model, containerId)
   }, [cellKey, containerId, layout, model])
-  const attrId = xAttrId && xAttrType === "numeric" ? xAttrId : yAttrId
+  const attrId = xAttrId && isNumericAttributeType(xAttrType) ? xAttrId : yAttrId
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/adornments/univariate-measures/box-plot/box-plot-adornment-component.tsx:52]: Agreed, same gap as the normal-curve thread: box-plot value/outlier/ICI labels go through `UnivariateMeasureAdornmentHelper.formatValueForScale` and will show epoch seconds on a date axis. Leaving open as a requested change.

File: v3/src/components/graph/adornments/movable-value/movable-value-adornment-component.scss [hunk 3]
```diff
@@ -9,7 +9,7 @@
   position: absolute;
   text-align: center;
   transform: translate(-50%, -55%);
-  width: 50px;
+  width: 65px;
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/adornments/movable-value/movable-value-adornment-component.scss]: Reasonable suggestion to consider — `white-space: nowrap` with a `min-width` would be more robust across date precisions/locales than a fixed 65px. Not blocking, but worth a look.

File: v3/src/components/graph/adornments/count/count-adornment-component.tsx [hunk 4]
```diff
@@ -31,6 +31,10 @@ export const CountAdornment = observer(function CountAdornment(props: IAdornment
   const { xScale, yScale } = useAdornmentAttributes()
   const dataConfig = useGraphDataConfigurationContext()
   const showMeasuresForSelection = !!dataConfig?.showMeasuresForSelection
+  // Observe the selection count so the component re-renders on selection change.
+  // The useEffect below recomputes count values on every render but is not itself reactive,
+  // so without this read the per-cell counts go stale when the user selects/deselects cases.
+  const _selectionCount = showMeasuresForSelection ? dataConfig?.selection.length ?? 0 : 0
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/adornments/count/count-adornment-component.tsx:37]: Fair point. These adornment components are D3/SVG-heavy and currently only model/handler-tested, so a component render test is a larger lift than this focused follow-up warrants. Existing model/handler tests pass. Not blocking on this.

File: v3/src/components/graph/adornments/movable-value/movable-value-adornment-component.tsx [hunk 5]
```diff
@@ -84,7 +86,11 @@ export const MovableValueAdornment = observer(function MovableValueAdornment(pro
   const refreshValue = useCallback((value: number, valueObjIndex: number) => {
     if (!value || !valueObjects.current[valueObjIndex]) return
     const multiScale = isVertical.current ? layout.getAxisMultiScale("bottom") : layout.getAxisMultiScale("left")
-    const displayValue = multiScale ? multiScale.formatValueForScale(value) : valueLabelString(value)
+    const primaryAxis = isVertical.current ? xAxis : yAxis
+    const dateAxis: IDateAxisModel | undefined = isDateAxisModel(primaryAxis) ? primaryAxis : undefined
+    const displayValue = multiScale
+      ? multiScale.formatValueForScale(value, !!dateAxis, dateAxis?.precisionForDisplay)
+      : valueLabelString(value)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/adornments/movable-value/movable-value-adornment-component.tsx:93]: Same as the count-adornment test thread — component-level render coverage for these D3/SVG adornments is out of scope for this focused follow-up. Not blocking.

File: v3/src/components/graph/adornments/count/count-adornment-component.tsx [hunk 6]
```diff
@@ -205,7 +209,7 @@ export const CountAdornment = observer(function CountAdornment(props: IAdornment
 
   useEffect(function respondToAxisDomainChange() {
     return mstReaction(
-      () => isNumericAxisModel(primaryAxisModel) ? primaryAxisModel?.domain.slice() : null,
+      () => isAnyNumericAxisModel(primaryAxisModel) ? primaryAxisModel?.domain.slice() : null,
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/adornments/count/count-adornment-component.tsx:212]: Same as the other test-coverage threads — not blocking this focused follow-up; component render tests for these adornments would be a separate effort.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Generally looks good -- Copilot identified and Claude confirmed some remaining gaps to fill. Details below in the following review developed in conjunction with Claude Code 🤖.

## PR Review Summary

**Changes:** 6 files, +31 / -14 lines

### What it does

Follow-up to CODAP-1320 / PR #2574. Four adornments (box plot, normal
curve, movable value, count) over-restrictively tested for `"numeric"`
attributes/axes and so misbehaved on date axes. This PR routes those
checks through the canonical `isNumericAttributeType` / `isAnyNumericAxisModel`
helpers so date axes are handled the same as numeric ones. Two extra
fixes found while testing: movable-value labels formatting raw
seconds-since-epoch on date axes, and per-cell count text going stale
on selection changes.

### The approach

- **Box plot / normal curve / movable value**: replaces hard-coded
  `xAttrType === "numeric"` with `isNumericAttributeType(xAttrType)`,
  affecting orientation, `attrId` selection, and drag math.
- **Count**: swaps `isNumericAxisModel` for `isAnyNumericAxisModel` in
  the `respondToAxisDomainChange` reaction so date-axis pan/zoom fires it.
- **Movable value labels**: passes `isDate` + `DateAxisModel.precisionForDisplay`
  to `multiScale.formatValueForScale(value, isDate, precision)` — matching
  the existing call pattern in `editable-slider-value.tsx`. SCSS label
  width widened 50px → 65px so dates fit on one line.
- **Count selection staleness**: adds `_selectionCount` read of
  `dataConfig.selection.length` in the render body so the `observer`
  re-renders on selection change.

### Assessment

The direction is right — it replaces ad-hoc string comparisons with the
project's canonical helpers, and the movable-value `formatValueForScale`
date arguments match the established slider call site. Verified locally:
lint clean on all 6 files, all 45 Jest tests in the four adornment
directories pass; CI (incl. full Cypress regression) is green.

However, the date-axis fix is incomplete: box-plot and normal-curve
**value labels** were not given the same date-formatting treatment that
movable-value labels received. Requesting changes to close that gap
before approval (see Issues).

### Issues

**Requested changes**

1. **Box-plot and normal-curve value labels render raw epoch seconds on
   date axes.** The PR enables date x-attributes for these two adornments
   (correct statistics, correct orientation), but their displayed values —
   box-plot value/outlier/ICI labels and normal-curve mean/stddev labels —
   still flow through `UnivariateMeasureAdornmentHelper.formatValueForScale`,
   which calls `MultiScale.formatValueForScale` with the default
   `isDate=false`. On a date axis these labels will show seconds-since-epoch
   instead of a date — exactly the bug this PR explicitly fixed for
   movable value. The helper needs the same `isDate` + precision treatment
   threaded through (likely by giving the helper access to the relevant
   date axis). (Copilot threads on `normal-curve-adornment-model.ts` and
   `box-plot-adornment-component.tsx` — left open.)

**Suggestion (non-blocking)**

2. The movable-value label uses a fixed `width: 65px` with no
   `white-space: nowrap`. A `nowrap` + `min-width` would be more robust
   across date precisions/locales. Author's call. (Copilot thread on
   `movable-value-adornment-component.scss` — left open.)

**Considered and dismissed**

- Copilot also asked for component-level render tests covering the
  selection-observation re-render, date-axis movable-value formatting,
  and date-axis count domain reaction. These adornment components are
  D3/SVG-heavy and currently only model/handler-tested; adding component
  render tests is a larger effort than this focused follow-up warrants.
  Acknowledged and not blocking (Copilot test-coverage threads resolved).
- `isNumericAttributeType` returns true for `"numeric"`, `"date"`, **and
  `"qualitative"`** — so these adornments now also treat qualitative
  x-attributes as vertical/primary, which the old `=== "numeric"` test
  excluded. Almost certainly harmless (these univariate adornments don't
  surface on qualitative plots) and using the canonical helper is the
  right call; noted, not blocking.

File: v3/src/components/graph/adornments/univariate-measures/normal-curve/normal-curve-adornment-model.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/adornments/univariate-measures/normal-curve/normal-curve-adornment-model.ts:133]: Addressed in 7d77697. UnivariateMeasureAdornmentHelper now receives xAxis/yAxis and threads isDate + DateAxisModel.precisionForDisplay into MultiScale.formatValueForScale, so the mean label renders as a date on a date axis.

Standard deviation needed an additional fix: SD is a *duration*, not an absolute date. Added formatDateDuration / formatDateDurationForScale in the same commit (reusing the seconds/minutes/hours/days/years thresholds from CODAP-1314's dateTimeSlopeUnit with new ScatterPlotModel.*Unit translation keys for plain "days"-style labels rather than slope's "per day"). SD now renders as e.g. "5.2 days".

File: v3/src/components/graph/adornments/univariate-measures/box-plot/box-plot-adornment-component.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/adornments/univariate-measures/box-plot/box-plot-adornment-component.tsx:52]: Addressed in 7d77697. Same helper change as the normal-curve thread: xAxis/yAxis flows through to the helper, so value/outlier/ICI labels on a date axis now format as dates. IQR (a duration) routes through the new formatDateDurationForScale, so it renders as e.g. "30 days" rather than a bogus date.

File: v3/src/components/graph/adornments/movable-value/movable-value-adornment-component.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (bfinzer) [reply v3/src/components/graph/adornments/movable-value/movable-value-adornment-component.scss]: Adopted in 7d77697 — switched to white-space: nowrap + min-width: 50px, dropped the fixed width.

File: v3/src/components/graph/utilities/graph-utils.ts [hunk 7]
```diff
@@ -331,6 +332,33 @@ function sigFigFractionDigits(value: number, sigFigs: number): number {
...
+}
+
+// Formats a duration (in seconds) using a human-scale time unit picked from the visible
+// date-axis range, e.g. "30 days" or "1 hour". Used for box-plot IQR, normal-curve
+// standard deviation, standard error, and similar duration-valued labels on date axes.
+export function formatDateDuration(durationSeconds: number, xRangeSeconds: number): string {
+  const { multiplier, unit } = dateTimeDurationUnit(xRangeSeconds)
+  const scaled = durationSeconds / multiplier
+  const formatted = formatValue(scaled, sigFigFractionDigits(scaled, 3))
+  return `${formatted} ${getDateUnitLabel(unit, scaled)}`
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/utilities/graph-utils.ts]: Good catch! Fixed in 18351034f. Pluralization now reads from the rounded display value (scaled rounded to the same fractionDigits used by formatValue), so 0.99998 → "1 second" instead of "1 seconds". Added an edge-case test that catches the bug.

**Review by kswenson — APPROVED** [review]
👍 Looks good -- added a few commits as part of the following review, developed in conjunction with Claude Code 🤖. See what you think.

## PR Re-Review Summary (commit `7d77697` on top of `e547896`)

**New commit:** `7d77697` — "CODAP-1328: format date-axis adornment labels as dates and durations"
**Changes:** 8 files, +87 / -18 lines

### What's new since the first review

Yesterday's review requested changes around date-axis label formatting for
box-plot and normal-curve adornments (their value/outlier/ICI/mean/SD labels
were still rendering raw epoch seconds because the helper called
`MultiScale.formatValueForScale` with the default `isDate=false`). This
follow-up commit threads `xAxis`/`yAxis` into `UnivariateMeasureAdornmentHelper`
and uses them for two purposes:

1. **Absolute dates** (mean, median, quartiles, whisker min/max, outliers,
   ICI bounds): `formatValueForScale` now reads `isDateAxisModel(primaryAxis)`
   and passes `isDate` + `dateAxis.precisionForDisplay` through to
   `MultiScale.formatValueForScale`. Same call shape as the existing
   `editable-slider-value.tsx` site.
2. **Durations** (IQR, normal-curve SD, standard error, MAD range): a new
   `formatDateDurationForScale` method calls the new
   `graph-utils.formatDateDuration(durationSeconds, xRangeSeconds)`, which
   picks a human-scale unit ("seconds", "minutes", "hours", "days", "years")
   off `dateTimeDurationUnit` — a sibling of the existing `dateTimeSlopeUnit`
   from CODAP-1314 sharing the same thresholds. After the reviewer's
   `da47f76b` follow-up, the unit label is sourced from the existing
   `V3.dateUnit.{unit}.{one|other}` keys via `getDateUnitLabel`, so it
   reuses already-translated strings and picks singular/plural correctly.

### How the requested changes were addressed

**1. Box-plot value/outlier/ICI labels render as dates on date axes — ✅ addressed.**
- `box-plot-adornment-component.tsx:413-420`: the 5 absolute stats (min,
  Q1, median, Q3, max) flow through `formatValueForScale`; IQR flows through
  the new `formatDateDurationForScale`. Both `lowerOutliers` and
  `upperOutliers` labels at lines 361/373 go through `formatValueForScale`
  (absolute), as do the ICI bounds at 469-470 (absolute). All correct.

**2. Normal-curve mean/SD labels render as date/duration on date axes — ✅ addressed.**
- `normal-curve-adornment-component.tsx:308-310`: mean uses
  `formatValueForScale` (absolute), SD uses `formatDateDurationForScale`
  (duration). Std-error string flows through
  `helper.computeTextContentForStdErr`, which now uses
  `formatDateDurationForScale` for `numStErrs * stdErr` (a half-range
  duration). Correct.

**3. SCSS suggestion (non-blocking from review #1) — ✅ adopted.**
- `movable-value-adornment-component.scss`: `width: 65px` replaced with
  `min-width: 50px` + `white-space: nowrap`. Matches the suggested pattern.

**4. Reviewer follow-up (commits `00e33671`, `da47f76b`, `18351034`).**
- `00e33671` — added a 5-branch test in `graph-utils.test.ts` parallel
  to the existing `dateTimeSlopeUnit` test, covering the
  seconds/minutes/hours/days/years thresholds and the value scaling.
- `da47f76b` — refactored `formatDateDuration` to use the existing
  `V3.dateUnit.{unit}.{one|other}` translation keys (already in
  en-US.json5 and de.json) via `getDateUnitLabel`, instead of the five
  duplicate `DG.ScatterPlotModel.*Unit` keys introduced earlier in this
  PR. This also picks up correct singular/plural forms via
  `Intl.PluralRules` (fixing a "1 seconds" pluralization bug at the
  exact-unit count). Added singular-form test cases parallel to the
  plural ones. Net delta: five duplicate translation keys removed.
- `18351034` — pluralize from the rounded display value rather than the
  raw `scaled` value, so e.g. `scaled=0.99998` (which displays as "1")
  yields "1 second" instead of "1 seconds". Caught by Copilot's
  re-review on `da47f76b`. Added boundary test cases at 0.99998 and 1.001.

### Implementation quality

- The constructor of `UnivariateMeasureAdornmentHelper` grew two optional
  positional params (`xAxis`, `yAxis`) after the existing optional
  `defaultLabelTopOffset`. All four call sites (box-plot, normal-curve,
  standard-error, simple) updated correctly; the three that don't need a
  `defaultLabelTopOffset` pass `undefined` and rely on the default param.
- The new `dateTimeDurationUnit` is a near-clone of `dateTimeSlopeUnit`
  (same thresholds, different label keys) — intentional, justified inline
  with a comment about CODAP-1314 reuse. Could in principle be DRY'd, but
  the labels semantically differ (slope = "per day", duration = "days"), so
  keeping them separate is reasonable.
- `formatDateDurationForScale` cleanly degrades to `formatValueForScale` on
  non-date axes, so callers can use it unconditionally for "is this a
  duration?" values without checking axis type first.
- New duration paths apply correctly to MAD's `displayRange` and standard
  error's `numStErrs * stdErr` — both are half-range deltas, not absolute
  values; comments explain why.

### CI

- Lint clean on all 8 files plus the test file. `npm run build:tsc` clean.
- 108 Jest tests across `univariate-measures`, `movable-value`, and `count`
  adornment directories all pass; the new `formatDateDuration` tests
  (plural + singular) pass alongside them.
- Full Cypress regression suite green on `18351034` (368 tests), as is
  every other CI check including Build/Lint/Jest/Smoke.
- `codecov/patch` had been failing at 67.19% on `7d77697` due to the new
  graph-utils helpers being uncovered; with the added tests it now passes.

### Assessment

The author's follow-up commit cleanly addresses both requested changes
and the SCSS suggestion from the first review. The implementation
correctly distinguishes "absolute date" values (which get
`isDate`+precision formatting) from "duration" values (which get
human-scale time units). The reviewer added a unit test for
`formatDateDuration`, refactored it to reuse the existing `V3.dateUnit`
translation key set with proper singular/plural handling, then fixed a
display/pluralization rounding mismatch caught by a fresh Copilot pass.
All adornment tests pass and CI is green on `18351034`. Ready to approve.

### Issues

No blocking issues. No outstanding suggestions.
